src/utils.cpp

```diff
@@ -334,6 +334,186 @@ void cnf_to_dimacs(Term cnf, std::ostringstream & y)
   }
 }
 
+
+
+
+
+Term term_gen(SmtSolver s, Sort boolsort, int& pt){
+  return (s->make_symbol("x"+std::to_string(pt++), boolsort));
+}
+
+
+  
+
+Term to_cnf(Term formula, SmtSolver s, Sort boolsort){
+  srand(time(NULL));
+  std::map<Term, Term>ma;//map ma contains the mapping of each term to it's new symbol in tseytin's transformation
+  
+  //returning symbolic constants directly
+  if(formula->is_symbolic_const()){
+    return formula;
+  }
+  //This vector is going to be broken down into it's symbolic constants, as we need to know which symbols have already been used so that we don't create symbols with the same name
+  
+
+
+  static int pt=1;// a pointer which will decide the next symbols name
+  
+  TermVec vec={formula};//vec stores the formulas which yet need to be given a symbolic name
+  //reduced stores (c) <-> (a op b). (c) as the first term of the pair and (a op b) as the second term in the pair
+  std::vector<std::pair<Term, Term>>reduced;
+  bool fir=0;//If fir is true it mean's we have named the parent formula i.e the formula given to us
+  Term parent;//parent will contain the symbolic name of the parent formula
+  while(!vec.empty()){
```

Comment:
No, I used the same idea as the earlier task(Of taking the last term, breaking it, and pushing the components back into the vector). Most of the code which I have written is regarding mapping to symbols, and for traversal, I have just written 5-6 lines. Should I use the walker for mapping?

Reply:
Yes I suggest you use the walker. Maybe it is just 5-6 lines now, but it might become more, and it is better to use an existing infrastructure rather than doing it again.

Reply:
If t exists in the form c1 op c2, I need to map these formulas to new symbols, so x1 would now be x2 op x3. But now I have to add the formula x1<->(x2 op x3) in tseitin's transformation. I think if I use a walker I would need to traverse the term again to pair up x1 with (x2 op x3) as just mapping the subformulas with new symbols wouldn't be enough. So, in my code, I have simultaneously mapped the subformulas to new symbols and added the pair to a vector.

Reply:
I see. I am wondering whether @makaimann has some idea here. @makaimann , @BhaskarGupta22 is implementing Tseitin transformation. I am trying to push to a walker-based implementation but it is not clear whether walkers are a right fit for this task. If you have an idea of how to use a walker for this, please let us know. Otherwise, we will go forward without using a walker.

Reply:
Sorry I didn't see this earlier. I might be missing something, but I think a walker would work fine for this. The walker does a pre- and post-order traversal. You would only be interested in the post-order traversal I believe. It would work something like this, create a new class that inherits from the `IdentityWalker`, and add whatever data structures you need (for example a mapping between terms and symbols). Then overload the virtual function `visit_term`. Within that function you would always update the cache if `preorder_` is false (i.e., it's in the post-order).

Then if you have something like `c1 op c2` that you want to name `x1`, you would first look up each of the arguments in the cache which would give you `x2 op x3`. Each of the children are guaranteed to have been processed already because you're in the post-order traversal. Then you could update the cache with `(c1 op c2) : x1`, and to a separate data structure save `x1 <-> (x2 op x3)`. 

Then you could query your walker object for those mappings (essentially your vector `reduced`) and create the clauses from there. All that being said, I'm not sure that it ends up being much less code because of the class boiler-plate. However, it automatically handles the DAG traversal, which you do manually by checking if it's already been mapped within `give_symbolic_name`.

Hopefully that's helpful. I think the walker might be cleaner, for similar reasons to my comment [here](https://github.com/makaimann/smt-switch/pull/252/files#r663207247)